src/relax/ir/block_builder.cc

```diff
@@ -29,44 +28,436 @@
+
+  /*!
+   * \brief lazily initialize function dedeup map.
+   */
+  void LazyInitCtxFuncDedupMap() {
+    if (ctx_func_dedup_map_ != nullptr) return;
+    ctx_func_dedup_map_ = std::make_unique<
+        std::unordered_map<BaseFunc, GlobalVar, StructuralHash, StructuralEqual>>();
+    for (const auto& kv : context_mod_->functions) {
+      const GlobalVar gv = kv.first;
+      const BaseFunc func = kv.second;
+      ctx_func_dedup_map_->emplace(func, gv);
+    }
+  }
+};
+
+//---------------------------------------
+// Normalization
+//---------------------------------------
 #define RELAX_EXPR_NORMALIZER_LEAF(OP) \
   Expr VisitExpr_(const OP* op) final { return GetRef<Expr>(op); }
 
-  RELAX_EXPR_NORMALIZER_LEAF(RuntimeDepShapeNode);
-  RELAX_EXPR_NORMALIZER_LEAF(ExternFuncNode);
-  RELAX_EXPR_NORMALIZER_LEAF(GlobalVarNode);
-  RELAX_EXPR_NORMALIZER_LEAF(OpNode);
-  RELAX_EXPR_NORMALIZER_LEAF(ShapeExprNode);
+// Invariance: If an Expr have a shape_ field, it must have already been normalized.
+class BlockBuilderImplWithNormalize : public BlockBuilderImpl,
+                                      private ExprFunctor<Expr(const Expr&)> {
+ public:
+  explicit BlockBuilderImplWithNormalize(IRModule context_mod) : BlockBuilderImpl(context_mod) {}
+
+  Expr Normalize(const Expr& expr) final {
+    Expr normalized = this->VisitExpr(expr);
+    // Invariance:
+    // After Normalize: an Expr always have checked_type (with the exception of Op).
+    if (!normalized->IsInstance<OpNode>()) {
+      ICHECK(normalized->checked_type_.defined())
+          << "The checked_type_ of an Expr except OpNode after "
+             "normalization must not be nullptr. However, this Expr does not have checked_type_: "
+          << normalized;
+    }
 
-  // TODO(@altanh): CopyOnWrite
+    return normalized;
+  }
 
-  Expr VisitExpr(const Expr& expr) {
-    // TODO(relax-team): generalize prim_func support
-    if (expr->IsInstance<tir::PrimFuncNode>()) {
-      return expr;
+  /*!
+   * \brief Normalize Argument values to call and other IR sub-fields.
+   * \param arg The argument.
+   * \return The normalized value.
+   *
+   * \note This function create a new binding for non-leaf expressions except for tuple.
+   */
+  Expr NormalizeArgument(Expr arg) {
+    if (!block_stack_.empty()) {
+      // cache lookup
+      BlockFrame* cur_frame = CurrentFrame();
+      auto it = cur_frame->normalize_binding_map.find(arg);
+      if (it != cur_frame->normalize_binding_map.end()) {
+        return it->second;
+      }
     }
-    Optional<Expr> post = expr_memo_.Get(expr);
-    if (post) {
-      ICHECK(post.as<VarNode>()) << "memoized expressions should map to variables";
-      return post.value();
+    // skip visit expr's cache, normalize arg
+    Expr post = ExprFunctor::VisitExpr(arg);
+
+    if (!IsLeafExpr(arg)) {
+      ICHECK(!block_stack_.empty()) << "Cannot normalize non-leaf without a scope";
+      Var var = this->Emit(post, "");
```

Comment:
We have a default value for the name string, so I think we can omit passing it explicitly here.
https://github.com/tlc-pack/relax/blob/6f669c15290e90efc5792da6a460817e24204281/include/tvm/relax/block_builder.h#L136-L145
```suggestion
      Var var = this->Emit(post);
```

Reply:
this is mainly to be explicit here, and somehow compiler can shadow the default arguments